src/util/interval.h

```diff
@@ -0,0 +1,472 @@
+  {
+  }
+};
+
+/// -∞ upper bound for intervals
+class min_exprt : public exprt
+{
+public:
+  explicit min_exprt(const typet &_type) : exprt(ID_min, _type)
+  {
+  }
+
+  explicit min_exprt(const exprt &_expr) : exprt(ID_min, _expr.type())
+  {
+  }
+};
+
+/// Represents an interval of values.
+/// Bounds should be constant expressions
+/// or min_exprt for the lower bound
+/// or max_exprt for the upper bound
+/// Also, lower bound should always be <= upper bound
+class constant_interval_exprt : public binary_exprt
+{
+public:
+  constant_interval_exprt(
+    const exprt &lower,
+    const exprt &upper,
+    const typet type)
+    : binary_exprt(lower, ID_constant_interval, upper, type)
+  {
+    PRECONDITION(is_well_formed());
+  }
+
+  constant_interval_exprt(const constant_interval_exprt &x)
+    : constant_interval_exprt(x.get_lower(), x.get_upper(), x.type())
+  {
+  }
+
+  explicit constant_interval_exprt(const exprt &x)
+    : constant_interval_exprt(x, x, x.type())
+  {
+  }
+
+  explicit constant_interval_exprt(const typet &type)
+    : constant_interval_exprt(min_exprt(type), max_exprt(type), type)
+  {
+  }
+
+  constant_interval_exprt(const exprt &lower, const exprt &upper)
+    : constant_interval_exprt(lower, upper, lower.type())
+  {
+  }
+
+  bool is_well_formed() const
+  {
+    bool b = true;
+
+    const typet &type = this->type();
+    const exprt &lower = get_lower();
+    const exprt &upper = get_upper();
+
+    b &= is_numeric() || type.id() == ID_bool || type.is_nil();
+
+    b &= type == lower.type();
+    b &= type == upper.type();
+
+    b &= is_valid_bound(lower);
+    b &= is_valid_bound(upper);
+
+    b &= !is_numeric() || is_bottom() || less_than_or_equal(lower, upper);
+
+    return b;
+  }
+
+  bool is_valid_bound(const exprt &expr) const
+  {
+    const irep_idt &id = expr.id();
+
+    bool b = true;
+
+    b &= id == ID_constant || id == ID_min || id == ID_max;
+
+    if(type().id() == ID_bool && id == ID_constant)
+    {
+      b &= expr == true_exprt() || expr == false_exprt();
+    }
+
+    return b;
+  }
+
+  static constant_interval_exprt tvt_to_interval(const tvt &val);
+
+  /* Naming scheme
+   *      is_[X]?  Returns bool / tvt
+   *      get_[X]? Returns relevant object
+   *      [X]      Generator of some object.
+   *      generate_[X]  generator used for evaluation
+   */
+
+  /* Getters */
+  const exprt &get_lower() const;
+  const exprt &get_upper() const;
+
+  /** SET OF ARITHMETIC OPERATORS */
+  constant_interval_exprt
+  handle_constant_unary_expression(const irep_idt &op) const;
+  constant_interval_exprt handle_constant_binary_expression(
+    const constant_interval_exprt &other,
+    const irep_idt &) const;
+
+  constant_interval_exprt eval(const irep_idt &unary_operator) const;
+  constant_interval_exprt
+  eval(const irep_idt &binary_operator, const constant_interval_exprt &o) const;
+
+  /* Unary arithmetic */
+  constant_interval_exprt unary_plus() const;
```

Comment:
Slightly more natural name
```suggestion
  constant_interval_exprt increment() const;
```

Reply:
nope - this is the + operator, as in `+x` not `x++` which is a no-op